src/node_jsrtapi.cc

```diff
@@ -126,51 +127,100 @@ namespace v8impl {
-public:
-  ObjectWrapWrapper(napi_value jsObject, void* nativeObj, napi_destruct destructor) {
-    _destructor = destructor;
-    _nativeObj = nativeObj;
+  // Adapter for JSRT external data + finalize callback.
+  class ExternalData {
+  public:
+    ExternalData(void* data, napi_finalize finalize_cb) : _data(data), _cb(finalize_cb) {
+    }
 
-    ObjectWrapWrapper::Wrap(jsObject, this);
+    void* Data() {
+      return _data;
+    }
 
-    JsSetObjectBeforeCollectCallback(reinterpret_cast<JsRef>(jsObject), this, JsObjectWrapWrapperBeforeCollectCallback);
-  }
+    // JsFinalizeCallback
+    static void CALLBACK Finalize(void* callbackState) {
+      ExternalData* externalData = reinterpret_cast<ExternalData*>(callbackState);
+      if (externalData != nullptr) {
+        if (externalData->_cb != nullptr) {
+          externalData->_cb(externalData->_data);
+        }
 
-  void* getValue() {
-    return _nativeObj;
-  }
+        delete externalData;
+      }
+    }
 
-  static void Wrap(napi_value jsObject, void* externalData) {
-    JsSetExternalData(jsObject, externalData);
-  }
+  private:
+    void* _data;
+    napi_finalize _cb;
+  };
 
-  static void* Unwrap(napi_value jsObject) {
-    ObjectWrapWrapper* wrapper = nullptr;
+  // Adapter for JSRT external callback + callback data.
+  class ExternalCallback {
+  public:
+    ExternalCallback(napi_callback cb, void* data) : _cb(cb), _data(data) {
+    }
 
-    JsGetExternalData(reinterpret_cast<JsValueRef>(jsObject), reinterpret_cast<void**>(&wrapper));
+    // JsNativeFunction
+    static JsValueRef CALLBACK Callback(JsValueRef callee,
+                                        bool isConstructCall,
+                                        JsValueRef *arguments,
+                                        unsigned short argumentCount,
+                                        void *callbackState) {
+      jsrtimpl::ExternalCallback* externalCallback =
+        reinterpret_cast<jsrtimpl::ExternalCallback*>(callbackState);
+
+      JsErrorCode error = JsNoError;
+      JsValueRef undefinedValue;
+      error = JsGetUndefinedValue(&undefinedValue);
+
+      CallbackInfo cbInfo;
+      cbInfo.thisArg = reinterpret_cast<napi_value>(arguments[0]);
+      cbInfo.isConstructCall = isConstructCall;
+
+      if (isConstructCall) {
+        // For constructor callbacks, replace the the 'this' arg with a new external object,
+        // to support wrapping a native object in the external object.
+        JsValueRef externalThis;
+        if (JsNoError == JsCreateExternalObject(
+          nullptr, jsrtimpl::ExternalData::Finalize, &externalThis)) {
+
+          // Copy the prototype from the default 'this' arg to the new 'external' this arg.
+          if (arguments[0] != nullptr) {
+            JsValueRef thisPrototype;
+            if (JsNoError == JsGetPrototype(arguments[0], &thisPrototype)) {
+              JsSetPrototype(externalThis, thisPrototype);
+            }
```

Comment:
We could throw a JavaScript exception if it failed here, but it didn't seem necessary.

The constructed object will just have the default Object prototype. In most cases it won't matter unless someone does an `instanceof` check. (Wrapped classes cannot extend JS classes.)